src/optlang/coinor_cbc_interface.py

```diff
@@ -0,0 +1,816 @@
+
+
+@six.add_metaclass(inheritdocstring)
+class Constraint(interface.Constraint):
+    _INDICATOR_CONSTRAINT_SUPPORT = False
+
+    def __init__(self, expression, sloppy=False, *args, **kwargs):
+        self._changed_expression = {}
+        super(Constraint, self).__init__(expression, sloppy=sloppy, *args, **kwargs)
+
+    def constraint_name(self, is_lb):
+        if is_lb:
+            return 'c_' + self.name + '_lower'
+        return 'c_' + self.name + '_upper'
+
+    @property
+    def primal(self):
+        if getattr(self, 'problem', None) is not None \
+           and self.problem.status == interface.OPTIMAL:
+            if self.lb is not None:
+                return self.problem._constr_primal(self, True)
+            if self.ub is not None:
+                return self.problem._constr_primal(self, False)
+        return None
+
+    @property
+    def dual(self):
+        if getattr(self, 'problem', None) is None:
+            return None
+        if self.lb is not None:
+            return self.problem._constr_dual(self, True)
+        if self.ub is not None:
+            return self.problem._constr_dual(self, False)
+        return None
+
+    def _update_bound(self, new, old, is_lb):
+        """Updates associated mip model with new constraint bounds."""
+        if getattr(self, 'problem', None) is None:
+            return
+
+        if old is None and new is not None:
+            self.problem._add_mip_constraint(self, is_lb)
+        elif new is None and old is not None:
+            self.problem._remove_mip_constraint(self, is_lb)
+        elif new is not None and old is not None:
+            self.problem._update_constraint_bound(self, is_lb)
+
+    @interface.Constraint.lb.setter
+    def lb(self, value):
+        self._check_valid_lower_bound(value)
+        self._lb, old_lb = value, getattr(self, '_lb', None)
+        self._update_bound(value, old_lb, True)
+
+    @interface.Constraint.ub.setter
+    def ub(self, value):
+        self._check_valid_upper_bound(value)
+        self._ub, old_ub = value, getattr(self, '_ub', None)
+        self._update_bound(value, old_ub, False)
+
+    @interface.Constraint.name.setter
+    def name(self, value):
+        super(Constraint, Constraint).name.fset(self, value)
+        if getattr(self, 'problem', None) is not None:
+            raise Exception('COIN-OR CBC doesn\'t support constraint name change')
+
+    def _get_expression(self):
+        if (self.problem is not None and self._changed_expression and
+            len(self.problem._variables) > 0):
+
+            coeffs = self._expression.as_coefficients_dict()
+
+            new_vars = set(self._changed_expression) - set(coeffs)
+
+            self._expression += symbolics.add([var * self._changed_expression[var]
+                                              for var in new_vars])
+
+            # Substitute var in expression with var * coef / old_coef
+            updates = [(var, var * coef / coeffs[var])
```

Comment:
Make updates a dictionary {var: var * coef / old_coeff} so it works with symengine and sympy.